rholang/src/main/scala/coop/rchain/rholang/interpreter/Interpreter.scala

```diff
@@ -25,36 +26,38 @@ object Interpreter {
 
   def apply[F[_]](implicit interpreter: Interpreter[F]): Interpreter[F] = interpreter
 
-  implicit def interpreter[F[_]](implicit F: Sync[F]): Interpreter[F] = new Interpreter[F] {
+  implicit def interpreter[F[_]](
+      implicit M: Monad[F],
+      S: Sync[F],
+      C: _cost[F],
+      E: _error[F]
+  ): Interpreter[F] =
+    new Interpreter[F] {
 
-    def evaluate(runtime: Runtime[F], term: String): F[EvaluateResult] =
-      evaluate(runtime, term, Cost.Max)
+      def evaluate(runtime: Runtime[F], term: String): F[EvaluateResult] =
+        evaluate(runtime, term, Cost.Max)
 
-    def evaluate(runtime: Runtime[F], term: String, initialPhlo: Cost): F[EvaluateResult] = {
-      implicit val rand: Blake2b512Random = Blake2b512Random(128)
-      for {
-        checkpoint <- runtime.space.createCheckpoint()
-        res        <- injAttempt(runtime.reducer, runtime.errorLog, term, initialPhlo)
-        _          <- if (res.errors.nonEmpty) runtime.space.reset(checkpoint.root) else F.unit
-      } yield res
-    }
+      def evaluate(runtime: Runtime[F], term: String, initialPhlo: Cost): F[EvaluateResult] = {
+        implicit val rand: Blake2b512Random = Blake2b512Random(128)
+        for {
+          checkpoint <- runtime.space.createCheckpoint()
+          res        <- injAttempt(runtime.reducer, runtime.errorLog, term, initialPhlo)
+          _          <- if (res.errors.nonEmpty) runtime.space.reset(checkpoint.root) else M.pure(())
+        } yield res
+      }
 
-    def injAttempt(
-        reducer: ChargingReducer[F],
-        errorLog: ErrorLog[F],
-        term: String,
-        initialPhlo: Cost
-    )(implicit rand: Blake2b512Random): F[EvaluateResult] = {
-      val parsingCost = accounting.parsingCost(term)
-      // TODO: charge(parsingCost) so that it is visible in the cost log
-      val phloAfterParsing = initialPhlo - parsingCost
-      if (phloAfterParsing.value <= 0)
-        EvaluateResult(parsingCost, Vector(OutOfPhlogistonsError)).pure[F]
-      else
+      def injAttempt(
```

Comment:
The initial (or any other) phlo should be never `set` in the first place - it should be provided upon construction of the cost account :) This takes place in Runtime.create, where rspace(s) and the dispatcher are created too. All those are re-used through node's lifetime though, while cost should have a much shorter lifecycle.

We've discussed this with @lukasz-golebiewski yesterday a bit, and my conclusion was that Runtime should only carry what's needed for a particular execution of the interperter (so just cost, most likely). Given what I've written above though, I now think we should just separate `_cost` from runtime and dispatcher, and pass it as a method parameter (until we reach a place where we have a `:_cost` requirement in the instance). So we'd pass `cost` explicitly to `doInj`, and then via implicit to the Interpreter (and ParBuilder?).

This way, we can get rid of the `set`-s without wreaking havoc around Runtime's lifecycle.

We should also try to make it just `FunctorTell` instead of a full `MonadState` (which is another refactoring long overdue).

Reply:
> The initial (or any other) phlo should be never set in the first place - it should be provided upon construction of the cost account :)

@ArturGajowy I was thinking the same thing at first. But, _cost instances belonging to the Runtime and system components span deploys. Therefore, you'll have to reset the balance of the account at the start of every deploy, which is here. Do you see the same thing?

Reply:
I do. And that's why we need to rip _cost out of runtime and its components, and only pass them via args when needed. We'll need to reset nothing - just throw away the old one, and create new one. The difference is substantial: just compare the powers of MonadState and FunctorTell.

Reply:
Doesn't the fact, that there's a `.set / .modify` method accessible wherever we have `_cost` give you shivers? With FunctorTell we'd only be doing `tell(costToCharge)`, and the instance would do the modification.